addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import hashlib
+import hmac
+import logging
+import re
+import time
+
+from datetime import datetime, timedelta
+
+from odoo import _, models
+from odoo.exceptions import AccessDenied
+from odoo.http import request
+
+from odoo.addons.auth_totp.models.totp import hotp, TOTP
+
+_logger = logging.getLogger(__name__)
+
+TOTP_RATE_LIMITS = {
+    'send_email': (_('You reached the limit of authentication mails sent for your account'), 10, 3600),
+    'code_check': (_('You reached the limit of code verifications for your account'), 10, 3600),
+}
+
+
+class Users(models.Model):
+    _inherit = 'res.users'
+
+    def _mfa_url(self):
+        r = super()._mfa_url()
+        if r is not None:
+            return r
+        ICP = self.env['ir.config_parameter'].sudo()
+        otp_required = False
+        if ICP.get_param('auth_totp.policy') == 'all_required':
+            otp_required = True
+        elif ICP.get_param('auth_totp.policy') == 'employee_required' and self.env.user.has_groups('base.group_user'):
+            otp_required = True
+        if otp_required:
+            return '/web/login/totp'
+
+    def _totp_check(self, code):
+        self._totp_rate_limit('code_check')
+        user = self.sudo()
+        if user.totp_enabled:
+            return super()._totp_check(code)
+
+        key = self._get_totp_mail_key()
+        match = TOTP(key).match(code, window=3600, timestep=3600)
```

Comment:
Note that contrary to a normal TOTP check where we're comparing a code generated by the user's device (with its own clock), here we're checking a code we generated ourselves. This means checking the "+1" window isn't very relevant, as we're unlikely to receive a code we generated in the future.

On the other hand it shouldn't hurt, and it may help for load-balanced servers where clock drift happens between the machines. Perhaps worth a small comment :)

Reply:
okidoki